### PR TITLE
Subgraph extraction in ONNX models

### PR DESCRIPTION
**Pull request by tomdol**
Related ticket number: 45414

### CODE REVIEW

**Comment by tomdol** [comment]
The subgraph extraction flow:
```
ONNXModelEditor::cut_graph_fragment
    SubgraphExtractor::add_new_inputs
    SubgraphExtractor::add_new_outputs
    SubgraphExtractor::extract_subgraph
        SubgraphExtractor::discover_output_contributors // called for each output
        SubgraphExtractor::extract_subgraph_from_onnx_model
```

File: ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp
Comment (t-jankowski) [inline ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp:388]: According to `discard_by_name` description its not meant to be used with outputs.

File: ngraph/test/util/onnx_test_util.cpp
Comment (pelszkow) [inline ngraph/test/util/onnx_test_util.cpp:67]: **Personally** I dislike this approach. 
My suggestion:
```
struct Result {
    bool valid{true};
    std::string message;
    static Result ok() { return {}; }
    static Result error(std::message m){return Resutl{false, std::move(m);}
}
```
Then you can check:
```
const auto result = func();
if ( !result.valid ) { std::cout << resutl.message << std::endl; }
```
I dream to have such class in some general scope.

File: ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp
Comment (pelszkow) [inline ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp:232]: Why `all_items` have to be `remove_if` several times?

**Review by pelszkow — COMMENTED** [review]
TBC

File: ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp
Comment (tomdol) [reply ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp:46]: I agree in general but in this particular file the lifetime of those names is guaranteed. The referenced values come from the top-level method invocation `ONNXModelEditor::cut_graph_fragment` which IMO makes it safe.

File: ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp
Comment (tomdol) [reply ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp:50]: This tag dispatch is not needed any more, I'll just remove this overload.

File: ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp
Comment (tomdol) [reply ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp:66]: Can't use ADL because protobuf does not provide the std::begin nor std::end overload for their `RepeatedPtrField` class. I'll switch to free functions instead of method calls though.

File: ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp
Comment (tomdol) [reply ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp:232]: This is a leftover from a previous approach, thanks for spotting it. I'll change it to a single call of remove_if.

File: ngraph/test/util/onnx_test_util.cpp
Comment (tomdol) [reply ngraph/test/util/onnx_test_util.cpp]: Good idea, done.

File: ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp
Comment (pelszkow) [reply ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp]: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c165-use-using-for-customization-points

File: ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp
Comment (tomdol) [reply ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp:377]: Good catch, fixed the docs.

**Review by t-jankowski — APPROVED** [review]

File: ngraph/frontend/onnx_import/include/onnx_import/editor/detail/subgraph_extraction.hpp
Comment (mbencer) [inline ngraph/frontend/onnx_import/include/onnx_import/editor/detail/subgraph_extraction.hpp:37]: Maybe just to clarify is worth to emphasize that nodes numeration start from 0 (from inputs to outputs in topological order) - if I understand correctly

**Review by mbencer — APPROVED** [review]
Very readable and well-tested implementation IMO. Great

File: ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp
Comment (tomdol) [reply ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp:131]: Good point but no. Here we create a new input from an existing initializer and initializers have static shapes. That's why the new input will also be statically shaped.

Comment (mbencer) [reply ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp:131]: ^^ OK, right ;)

File: ngraph/frontend/onnx_import/src/editor/editor.cpp
Comment (mbencer) [inline ngraph/frontend/onnx_import/src/editor/editor.cpp:247]: You can use here `OutputEdge` alias

Comment (tomdol) [reply ngraph/frontend/onnx_import/src/editor/editor.cpp:247]: Good catch, it was supposed to be a strong type alias and that way this method wouldn't even compile. I'll update the code to address it.

**Review by jiwaszki — APPROVED** [review]

File: ngraph/frontend/onnx_import/include/onnx_import/editor/editor.hpp
Comment (ilyachur) [inline ngraph/frontend/onnx_import/include/onnx_import/editor/editor.hpp:94]: Does it means that we can remove import_network method for `ONNXEditor`?

Comment (tomdol) [reply ngraph/frontend/onnx_import/include/onnx_import/editor/editor.hpp:94]: Perhaps we should go this way but let me discuss it with the team today and get back to you.

Comment (tomdol) [reply ngraph/frontend/onnx_import/include/onnx_import/editor/editor.hpp:94]: This will be benchmarked and addressed in a separate PR where the editor will be extracted to a standalone library.

File: ngraph/frontend/onnx_import/src/utils/onnx_test_util.hpp
Comment (ilyachur) [inline ngraph/frontend/onnx_import/src/utils/onnx_test_util.hpp:27]: Looks like we don't need to export this struct.

File: ngraph/frontend/onnx_import/src/utils/onnx_test_util.hpp
Comment (ilyachur) [inline ngraph/frontend/onnx_import/src/utils/onnx_test_util.hpp:51]: Why do we need to have these struct and method inside the ONNX importer if we use their only for test purposes?

Comment (tomdol) [reply ngraph/frontend/onnx_import/src/utils/onnx_test_util.hpp:51]: This was moved here to avoid protobuf related problems. When the editor is extracted to a separate library this can perhaps be moved somewhere else.

File: ngraph/frontend/onnx_import/src/utils/onnx_test_util.hpp
Comment (tomdol) [reply ngraph/frontend/onnx_import/src/utils/onnx_test_util.hpp:27]: For now we have to so that it can be used in the editor tests.

File: ngraph/frontend/onnx_import/src/utils/onnx_test_util.hpp
Comment (ilyachur) [reply ngraph/frontend/onnx_import/src/utils/onnx_test_util.hpp:51]: @tomdol Can we try to avoid such code? Maybe we can make a call and discuss it, because for me looks strange that classes and methods which are used only in the tests, we need to export and have in the core library.

Comment (tomdol) [reply ngraph/frontend/onnx_import/src/utils/onnx_test_util.hpp:51]: Sure, we can sync up about it. But just FYI @mbencer is working on the extraction of everything editor-related to a separate library so he will move this code too.

Comment (tomdol) [reply ngraph/frontend/onnx_import/src/utils/onnx_test_util.hpp]: As discussed, I've temporarily switched to comparison of binary strings containing the extracted models. I've also created a ticket to address it after the editor becomes a separate library: 49339

File: ngraph/test/models/onnx/model_editor/reference/subgraph__initializer_without_matching_input_tail_cut.prototxt
Comment (ilyachur) [inline ngraph/test/models/onnx/model_editor/reference/subgraph__initializer_without_matching_input_tail_cut.prototxt:2]: @tomdol Why do we need to have this field?

File: ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp
Comment (tomdol) [reply ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp:71]: We're iterating over a topological order of nodes. This means that the node producing input to a given node has to be located before. This way we can start looking at the position `current_node_idx - 1` and go towards the beginning of the container.

File: ngraph/test/models/onnx/model_editor/reference/subgraph__initializer_without_matching_input_tail_cut.prototxt
Comment (tomdol) [reply ngraph/test/models/onnx/model_editor/reference/subgraph__initializer_without_matching_input_tail_cut.prototxt:2]: We don't, it's optional. I can remove this field but it will require a full run of CI.

File: ngraph/frontend/onnx_import/include/onnx_import/editor/detail/subgraph_extraction.hpp
Comment (slyalin) [inline ngraph/frontend/onnx_import/include/onnx_import/editor/detail/subgraph_extraction.hpp:50]: How can one obtain a valid value for m_node_idx? Briefly looking in the code I see that a valid value is really required but there is no way for use to deduce this index based on some user-visible name without parsing model proto file. Even Netron doesn't give this index.

Can we make it at least optional? In this case it would mean cutting by a tensor name and providing a single input for all the consumers.

File: ngraph/frontend/onnx_import/include/onnx_import/editor/detail/subgraph_extraction.hpp
Comment (slyalin) [inline ngraph/frontend/onnx_import/include/onnx_import/editor/detail/subgraph_extraction.hpp:69]: Assuming that `node 5` is an operation and `(in_A)` etc. are tensors, this documentation doesn't provide rationale why we are specifying node index, because there is only one consumer for each of the tensors on the schematics. In this case the tensor name should be enough to specify the cutting point.

Comment (tomdol) [reply ngraph/frontend/onnx_import/include/onnx_import/editor/detail/subgraph_extraction.hpp:69]: As discussed offline, this will be addressed in a separate ticket: 47578

File: ngraph/frontend/onnx_import/include/onnx_import/editor/detail/subgraph_extraction.hpp
Comment (tomdol) [reply ngraph/frontend/onnx_import/include/onnx_import/editor/detail/subgraph_extraction.hpp:50]: Similarily to the comment below - TBD in 47578

**Comment by tomdol** [comment]
@openvinotoolkit/openvino-ngraph-maintainers all feedback addressed. Can we merge it or should we wait for anyone else's review?

**Comment by ilyachur** [comment]
@tomdol I propose to wait for release branch creation. Because the feature freeze happened and we don't want to have this feature in the 2021.3 release.
So my suggestion is to wait for release branch creation.

**Review by tsocha — APPROVED** [review]

**Review by postrational — APPROVED** [review]

**Pull request merged**
